Sources/SourceControl/RepositoryManager.swift

```diff
@@ -186,70 +155,48 @@ public class RepositoryManager {
                 self.pendingLookups[repository] = group
                 self.pendingLookupsLock.unlock()
 
-                let subpath = RelativePath(repository.fileSystemIdentifier)
-                let handle = RepositoryHandle(manager: self, repository: repository, subpath: subpath)
-                let repositoryPath = self.path.appending(handle.subpath)
-
-                // Change the state to pending.
-                // Make sure destination is free.
-                try? self.fileSystem.removeFileTree(repositoryPath)
+                // inform delegate that we are starting to fetch
+                // calculate if cached (for delegate call) outside queue as it may change while queue is processing
                 let isCached = self.cachePath.map{ self.fileSystem.exists($0.appending(handle.subpath)) } ?? false
-
-                // Inform delegate.
-                queue.async {
+                delegateQueue.async {
                     let details = FetchDetails(fromCache: isCached, updatedCache: false)
-                    self.delegate?.fetchingWillBegin(handle: handle, fetchDetails: details)
+                    self.delegate?.willFetch(repository: handle.repository, details: details)
                 }
 
-                let result: LookupResult
                 let start = DispatchTime.now()
+                let lookupResult: Result<RepositoryHandle, Error>
+                let delegateResult: Result<FetchDetails, Error>
 
-                // Fetch the repo.
-                var fetchError: Swift.Error? = nil
-                var fetchDetails: FetchDetails? = nil
                 do {
-                    // Start fetching.
-                    fetchDetails = try self.fetchAndPopulateCache(handle: handle, repositoryPath: repositoryPath, delegateQueue: queue)
```

Comment:
This is great — I think the reordering here addresses a bug where the `didFetch` callback could happen before an error was thrown about the repository, which at least one of the clients of libSwiftPM had to work around.